### PR TITLE
Migrate ruby to new importers

### PR DESCRIPTION
**Pull request by ziadhany**
Reference: #796
Signed-off-by: Ziad <ziadhany2016@gmail.com>

### CODE REVIEW

**Comment by ziadhany** [comment]
some of ruby logs.  Can you have a look at `categorize_versions`, I am not sure if am made the right implementation for it .

```
Invalid Rubygems Version 1.13.2 , Invalid constraints sequence: [VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6')), VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6.2')), VersionConstraint(comparator='<', version=RubygemsVersion(string='5.3'))]
Invalid Rubygems Version 1.13.1 , Invalid constraints sequence: [VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6')), VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6.2')), VersionConstraint(comparator='<', version=RubygemsVersion(string='5.3'))]
Invalid Rubygems Version 1.13.0 , Invalid constraints sequence: [VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6')), VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6.2')), VersionConstraint(comparator='<', version=RubygemsVersion(string='5.3'))]
Invalid Rubygems Version 1.12.5 , Invalid constraints sequence: [VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6')), VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6.2')), VersionConstraint(comparator='<', version=RubygemsVersion(string='5.3'))]
Invalid Rubygems Version 1.12.4 , Invalid constraints sequence: [VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6')), VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6.2')), VersionConstraint(comparator='<', version=RubygemsVersion(string='5.3'))]
Invalid Rubygems Version 1.12.3 , Invalid constraints sequence: [VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6')), VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6.2')), VersionConstraint(comparator='<', version=RubygemsVersion(string='5.3'))]
Invalid Rubygems Version 1.12.2 , Invalid constraints sequence: [VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6')), VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6.2')), VersionConstraint(comparator='<', version=RubygemsVersion(string='5.3'))]
Invalid Rubygems Version 1.12.1 , Invalid constraints sequence: [VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6')), VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6.2')), VersionConstraint(comparator='<', version=RubygemsVersion(string='5.3'))]
Invalid Rubygems Version 1.12.0 , Invalid constraints sequence: [VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6')), VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6.2')), VersionConstraint(comparator='<', version=RubygemsVersion(string='5.3'))]
Invalid Rubygems Version 1.11.2 , Invalid constraints sequence: [VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6')), VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6.2')), VersionConstraint(comparator='<', version=RubygemsVersion(string='5.3'))]
Invalid Rubygems Version 1.11.1 , Invalid constraints sequence: [VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6')), VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6.2')), VersionConstraint(comparator='<', version=RubygemsVersion(string='5.3'))]
Invalid Rubygems Version 1.11.0 , Invalid constraints sequence: [VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6')), VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6.2')), VersionConstraint(comparator='<', version=RubygemsVersion(string='5.3'))]
Invalid Rubygems Version 1.10.2 , Invalid constraints sequence: [VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6')), VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6.2')), VersionConstraint(comparator='<', version=RubygemsVersion(string='5.3'))]
Invalid Rubygems Version 1.10.1 , Invalid constraints sequence: [VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6')), VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6.2')), VersionConstraint(comparator='<', version=RubygemsVersion(string='5.3'))]
Invalid Rubygems Version 1.9.1 , Invalid constraints sequence: [VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6')), VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6.2')), VersionConstraint(comparator='<', version=RubygemsVersion(string='5.3'))]
Invalid Rubygems Version 1.9.0 , Invalid constraints sequence: [VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6')), VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6.2')), VersionConstraint(comparator='<', version=RubygemsVersion(string='5.3'))]
Invalid Rubygems Version 1.8.1 , Invalid constraints sequence: [VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6')), VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6.2')), VersionConstraint(comparator='<', version=RubygemsVersion(string='5.3'))]
Invalid Rubygems Version 1.8.0 , Invalid constraints sequence: [VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6')), VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6.2')), VersionConstraint(comparator='<', version=RubygemsVersion(string='5.3'))]
Invalid Rubygems Version 1.7.0 , Invalid constraints sequence: [VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6')), VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6.2')), VersionConstraint(comparator='<', version=RubygemsVersion(string='5.3'))]
Invalid Rubygems Version 1.6.0 , Invalid constraints sequence: [VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6')), VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6.2')), VersionConstraint(comparator='<', version=RubygemsVersion(string='5.3'))]
Invalid Rubygems Version 1.5.1 , Invalid constraints sequence: [VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6')), VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6.2')), VersionConstraint(comparator='<', version=RubygemsVersion(string='5.3'))]
Invalid Rubygems Version 1.5.0 , Invalid constraints sequence: [VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6')), VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6.2')), VersionConstraint(comparator='<', version=RubygemsVersion(string='5.3'))]
Invalid Rubygems Version 1.4.0 , Invalid constraints sequence: [VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6')), VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6.2')), VersionConstraint(comparator='<', version=RubygemsVersion(string='5.3'))]
Invalid Rubygems Version 1.3.1 , Invalid constraints sequence: [VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6')), VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6.2')), VersionConstraint(comparator='<', version=RubygemsVersion(string='5.3'))]
Invalid Rubygems Version 1.3.0 , Invalid constraints sequence: [VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6')), VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6.2')), VersionConstraint(comparator='<', version=RubygemsVersion(string='5.3'))]
Invalid Rubygems Version 1.2.0 , Invalid constraints sequence: [VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6')), VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6.2')), VersionConstraint(comparator='<', version=RubygemsVersion(string='5.3'))]
Invalid Rubygems Version 1.1.0 , Invalid constraints sequence: [VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6')), VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6.2')), VersionConstraint(comparator='<', version=RubygemsVersion(string='5.3'))]
Invalid Rubygems Version 1.0.1 , Invalid constraints sequence: [VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6')), VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6.2')), VersionConstraint(comparator='<', version=RubygemsVersion(string='5.3'))]
Invalid Rubygems Version 1.0.0 , Invalid constraints sequence: [VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6')), VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6.2')), VersionConstraint(comparator='<', version=RubygemsVersion(string='5.3'))]
Invalid Rubygems Version 0.9.5 , Invalid constraints sequence: [VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6')), VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6.2')), VersionConstraint(comparator='<', version=RubygemsVersion(string='5.3'))]
Invalid Rubygems Version 0.9.0 , Invalid constraints sequence: [VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6')), VersionConstraint(comparator='>=', version=RubygemsVersion(string='5.2.6.2')), VersionConstraint(comparator='<', version=RubygemsVersion(string='5.3'))]
Successfully imported data using vulnerabilities.importers.ruby.RubyImporter
```

File: vulnerabilities/importers/ruby.py [hunk 1]
```diff
@@ -7,132 +7,196 @@
...
+    spdx_license_expression = "unknow"
+    config = GitConfig(
+        repository_url="https://github.com/rubysec/ruby-advisory-db.git",
+        working_directory="ruby",
+        branch="master",
+    )
+    cutoff_timestamp = 1
+
+    def __init__(self):
+        super().__init__(config=self.config, cutoff_timestamp=self.cutoff_timestamp)
```
Comment (TG1999) [inline vulnerabilities/importers/ruby.py:45]: Version API needs to be handled at improver side, please check github and gitlab importer for same

File: vulnerabilities/importers/ruby.py [hunk 2]
```diff
@@ -7,132 +7,196 @@
...
+        self.pkg_manager_api = RubyVersionAPI()
```
Comment (TG1999) [inline vulnerabilities/importers/ruby.py:46]: @ziadhany this is not migrated correctly, API code should not go in the importer, improver should have dealt with it. This doesn't adhere to the new structure of importer-improver and looks mostly as the same old structure.

**Comment by TG1999** [comment]
@ziadhany what's the status on this ?

**Comment by TG1999** [comment]
Dependent on https://github.com/nexB/univers/pull/92/files

**Comment by TG1999** [comment]
@ziadhany invert functionality has been added to univers, please use https://pypi.org/project/univers/30.9.1/

**Comment by TG1999** [comment]
@ziadhany gentle ping

**Comment by ziadhany** [comment]
Ruby importer and improver logs : 
[ruby_imp_logs.zip](https://github.com/nexB/vulnerablecode/files/10548210/ruby_imp_logs.zip)

**Review by TG1999 — COMMENTED** [review]
@ziadhany Thanks++, some review comments for your consideration, please rebase your branch as well.

File: vulnerabilities/importers/ruby.py [hunk 3]
```diff
@@ -7,132 +7,146 @@
...
+
+def parse_ruby_advisory(record, schema_type):
+    """
+    Parse a ruby advisory file and return an AdvisoryData or None.
+    Each advisory file contains the advisory information in YAML format.
+    Schema: https://github.com/rubysec/ruby-advisory-db/tree/master/spec/schemas
+    """
+    if schema_type == "gems":
+        package_name = record.get("gem")
+        library = record.get("library")  # not used
```
[the remark this answers is not in the document]
Comment (ziadhany) [reply vulnerabilities/importers/ruby.py:56]: I thought maybe we can make use of it in the future,  ( ok I will remove it )

**Comment by ziadhany** [comment]
logs : [importer--improver-ruby.zip](https://github.com/nexB/vulnerablecode/files/12458531/importer--improver-ruby.zip)

File: vulnerabilities/importers/ruby.py [hunk 4]
```diff
@@ -7,132 +7,146 @@
...
 from vulnerabilities.utils import load_yaml
-from vulnerabilities.utils import nearest_patched_package
 
+logger = logging.getLogger(__name__)
+
+
+class RubyImporter(Importer):
+    license_url = "https://github.com/rubysec/ruby-advisory-db/blob/master/LICENSE.txt"
+    spdx_license_expression = "unknown"
+    repo_url = "git+https://github.com/rubysec/ruby-advisory-db"
```
[the remark this answers is not in the document]
Comment (ziadhany) [reply vulnerabilities/importers/ruby.py:34]: Done

**Review by TG1999 — APPROVED** [review]
LGTM!

File: vulnerabilities/importers/ruby.py [hunk 5]
```diff
@@ -7,132 +7,147 @@
...
+            date_published=get_publish_time(record),
         )
 
-    @staticmethod
-    def categorize_versions(all_versions, unaffected_version_ranges):
 
-        for id, elem in enumerate(unaffected_version_ranges):
-            unaffected_version_ranges[id] = VersionRange.from_scheme_version_spec_string(
-                "semver", elem
+def get_affected_packages(record, purl):
```
Comment (TG1999) [inline vulnerabilities/importers/ruby.py:82]: @ziadhany please add docstring

Comment (TG1999) [reply vulnerabilities/importers/ruby.py:82]: Also please add some unit tests for this function with some corner cases

File: vulnerabilities/importers/ruby.py [hunk 6]
```diff
@@ -7,132 +7,147 @@
...
+            aliases=get_aliases(record),
+            summary=get_summary(record),
+            affected_packages=get_affected_packages(record, purl),
+            references=get_references(record),
+            date_published=get_publish_time(record),
+        )
 
+    elif schema_type == "rubies":
+        engine = record.get("engine")  # engine enum: [jruby, rbx, ruby]
+        purl = PackageURL(type="ruby", name=engine)
```
Comment (TG1999) [inline vulnerabilities/importers/ruby.py:72]: what if record doesn't have "engine" property ?

File: vulnerabilities/importers/ruby.py [hunk 7]
```diff
@@ -7,132 +7,147 @@
...
+                    severities=[
+                        VulnerabilitySeverity(system=SCORING_SYSTEMS["cvssv3"], value=cvss_v3)
+                    ],
+                )
+            )
+    return references
+
+
+def get_publish_time(record):
+    return parse(record["date"]).replace(tzinfo=UTC)
```
Comment (TG1999) [inline vulnerabilities/importers/ruby.py:147]:
```suggestion
    date = record.get("date")
    if not date:
        return
    return parse(date).replace(tzinfo=UTC)
```

File: vulnerabilities/importers/ruby.py [hunk 8]
```diff
@@ -7,132 +7,147 @@
...
-                    safe_versions.append(i)
-                    is_vulnerable = True
-                    break
-
-            if not is_vulnerable:
-                vulnerable_versions.append(i)
-
-        return safe_versions, vulnerable_versions
+def get_summary(record):
+    title = record.get("title")
```
Comment (TG1999) [inline vulnerabilities/importers/ruby.py:151]:
```suggestion
    title = record.get("title") or ""
```

File: vulnerabilities/importers/ruby.py [hunk 9]
```diff
@@ -7,132 +7,147 @@
...
+    description = record.get("description", "")
```
[the remark this answers is not in the document]
Comment (ziadhany) [reply vulnerabilities/importers/ruby.py:152]: what is the difference between these 2 lines of code?

File: vulnerabilities/importers/ruby.py [hunk 10]
```diff
@@ -7,132 +7,147 @@
...
+
+
+def parse_ruby_advisory(record, schema_type):
+    """
+    Parse a ruby advisory file and return an AdvisoryData or None.
+    Each advisory file contains the advisory information in YAML format.
+    Schema: https://github.com/rubysec/ruby-advisory-db/tree/master/spec/schemas
+    """
+    if schema_type == "gems":
         package_name = record.get("gem")
```
[the remark this answers is not in the document]
Comment (ziadhany) [reply vulnerabilities/importers/ruby.py:59]: According to [ruby-advisory-db](https://github.com/rubysec/ruby-advisory-db/blob/master/spec/schemas/gem.yml#L6C14-L6C19), this field is required

File: vulnerabilities/importers/ruby.py (hunk 5 above)
Comment (ziadhany) [reply vulnerabilities/importers/ruby.py:89]: Done

File: vulnerabilities/importers/ruby.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (TG1999) [reply vulnerabilities/importers/ruby.py]: @ziadhany suppose, record actually have "description" in it, and that "description" property is `None`, then your variable will be assigned `None`.

**Comment by TG1999** [comment]
@ziadhany please rebase your PR

File: vulnerabilities/importers/ruby.py [hunk 11]
```diff
@@ -7,132 +7,162 @@
...
-            if data.get("gem"):
-                packages.add(data["gem"])
-
-        return packages
-
-    def process_file(self, path) -> List[AdvisoryData]:
-        record = load_yaml(path)
+class RubyImporter(Importer):
+    license_url = "https://github.com/rubysec/ruby-advisory-db/blob/master/LICENSE.txt"
+    spdx_license_expression = "unknown"
```
Comment (TG1999) [inline vulnerabilities/importers/ruby.py:33]: @ziadhany @pombredanne what should be done for this?

**Review by TG1999 — CHANGES_REQUESTED** [review]

File: vulnerabilities/importers/ruby.py (hunk 11 above)
Comment (TG1999) [reply vulnerabilities/importers/ruby.py:33]: @ziadhany let's get some stats, how much data we can ingest once we filter out the data that is associated with OSVDB ?

Comment (TG1999) [reply vulnerabilities/importers/ruby.py:33]: https://scancode-licensedb.aboutcode.org/public-domain-disclaimer.html , 
```
spdx_license_expression = "LicenseRef-scancode-public-domain-disclaimer"
notice = """
If you submit code or data to the ruby-advisory-db that is copyrighted by
yourself, upon submission you hereby agree to release it into the public
domain.

The data imported from the ruby-advisory-db have been filtered to exclude 
any non-public domain data from the data copyrighted by the Open 
Source Vulnerability Database (http://osvdb.org).

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
"""
```

File: vulnerabilities/importers/ruby.py [hunk 12]
```diff
@@ -7,132 +7,188 @@
...
-        for f in files:
-            data = load_yaml(f)
-            if data.get("gem"):
-                packages.add(data["gem"])
-
-        return packages
-
-    def process_file(self, path) -> List[AdvisoryData]:
-        record = load_yaml(path)
+class RubyImporter(Importer):
```
Comment (TG1999) [inline vulnerabilities/importers/ruby.py:32]: Please add this importer here in this test as well https://github.com/nexB/vulnerablecode/blob/398ca99432ebb1fc7dafd6c8d24e59321a074df6/vulnerabilities/tests/test_data_source.py#L114

**Review by TG1999 — CHANGES_REQUESTED** [review]
@ziadhany thanks++, minor nit for your consideration

**Review by TG1999 — APPROVED** [review]
LGTM! thanks please do the suggested changes and feel free to squash and merge this PR : )

File: vulnerabilities/importers/ruby.py [hunk 13]
```diff
@@ -7,132 +7,188 @@
...
-                "semver", elem
+def get_references(record) -> [Reference]:
+    references = []
+    cvss_v2 = record.get("cvss_v2")
+    cvss_v3 = record.get("cvss_v3")
+
+    if record.get("url"):
+        if not (cvss_v2 or cvss_v3):
+            references.append(Reference(url=record.get("url")))
+        if cvss_v2:
```
Comment (TG1999) [inline vulnerabilities/importers/ruby.py:163]: Do we want to keep CVSSV2 ? 

cc @pombredanne @DennisClark

Comment (ziadhany) [reply vulnerabilities/importers/ruby.py]: Done

**Comment by TG1999** [comment]
@ziadhany thanks for your effort and persistence : )
